### DISCUSSION TITLE
Redis Initial connection takes more than 1.5 seconds in aws Lambda function, wonder why?

### DISCUSSION THREAD

**Discussion by y2241** [redis/lettuce#2272] [General]
My app used to have 1.5 seconds as the connection timeout and 2 seconds as socket timeout. But our aws lambda function kept throwing Redis Initial connection timeout exception. After I bumped up the connection timeout to 3 seconds, and added bunch of logs, the exception stopped and I notices that many initial connection from lambda function took more than 1.5 seconds (but less than 1.7 seconds). 
The time is spent on this method inside the RedisClient.java:
public <K, V> StatefulRedisConnection<K, V> connect(RedisCodec<K, V> codec) {

        checkForRedisURI();

        return getConnection(connectStandaloneAsync(codec, this.redisURI, getDefaultTimeout()));
} 
Wonder why it could take this long to establish a connection. Our app only uses one redis client and the client just opens one connection (it does not use connection pool). I turned on the "io.lettuce.core" trace log and did not see anything major. The "Hello" command ran fast. 
Is this a lambda specific thing or the our redis on aws side is just slow (network?).

If you have some ideas or suggestions, please share with me. Thanks!

**Comment by mp911de** [comment]
Likely, the netty thread pool needs to start up. If you limit the I/O/worker thread pools to a single thread, then you should see a much better startup performance

**Reply by y2241** [reply]
Thanks mp911de for quick reply. Could you explain more on "If you limit the I/O/worker thread pools to a single thread"? I would like to give a quick try, but not sure how to set this "single thread".

**Reply by y2241** [reply]
Can I also piggy back another question here as it is part of our findings. We use scriptLoad api to cache some lua scripts to redis cache. We do this everytime from several places as we don't know if those scripts are already uploaded to redis. I hope we can only do it once. Is there way to tell if the scripts are already cached on redis side? 
The reason we run those lua scripts is because we implement our own RateLimiter using redis, so the simple increase count and decrease count and reset count have to be atomic and thread safe. I wonder if there any other way to accomplish this so that we don't have to run lua scripts on redis server side. Instead, just manage the count on client side using lettuce library?

**Reply by y2241** [reply]
Mark, we are using DefaultClientResources which is already like using 2 as MIN_IO_THREADS:
 int threads = Math.max(1, SystemPropertyUtil.getInt("io.netty.eventLoopThreads",
                Math.max(MIN_IO_THREADS, Runtime.getRuntime().availableProcessors())));
Do you think it will be benefit to use Math.max(1, Runtime.getRuntime().availableProcessors())))?
Also, we are using getRedisClient().connect() right now in the code. Do you think it is beneficial to use asynchronize connection such as getRedisClient().connect().async()?
